src/HypothesisTests.jl

```diff
@@ -69,34 +69,42 @@ If `tail` is `:both` (default), then the p-value for the two-sided test is retur
 function pvalue end
 
 # Basic function for finding a p-value given a distribution and tail
-pvalue(dist::ContinuousUnivariateDistribution, x::Number; tail=:both) =
+function pvalue(dist::ContinuousUnivariateDistribution, x::Number; tail=:both)
+    check_tail(tail)
+
     if tail == :both
-        min(2 * min(cdf(dist, x), ccdf(dist, x)), 1.0)
+        min(2 * min(cdf(dist, x), ccdf(dist, x)), 1)
```

Comment:
Why not keep this one? I would actually have done the contrary. :-D

Reply:
Because in this definition we don't restrict, and hence don't know about, the types of `x` and possible parameters of the distribution. So `cdf` and `ccdf` might not return a `Float64` but eg a `Float32`. And therefore the `Float64` literal can cause a type instability.

Reply:
Hm I thought `min` promotes the arguments but it seems it doesn't? https://github.com/JuliaLang/julia/blob/bf534986350a991e4a1b29126de0342ffd76205e/base/operators.jl#L506

Then we should explicitly promote or maybe just use `oneunit`.

Reply:
Yeah that was my concern. Maybe we should just add `::Float64` to all p-value methods? I don't see under which circumstances one would prefer getting a p-value in another type.

Reply:
I think the better approach is to promote here. For all standard arguments it will be `Float64`, so it won't cause any surprising return types.

Restricting everything to `Float64` seems unnecessary (most methods don't allow such general arguments and hence there's no such freedom as in this method here) and a bit unidiomatic - why restrict it if it works fine? More importantly though, I learnt in Distributions how problematic for e.g. GPU support and AD a `Float64`-only approach and how long it takes to (somewhat) fix it. So on purpose, both in StatsFuns and Distributions neither arguments nor return values of `cdf` and `ccdf` (which are called here) are restricted to `Float64`. Thus even though I don't have a more concrete example at hand apart from abstract GPU and AD applications, I think it's better to just fix the line here.

Reply:
On a second thought, is `min(..., 1)` actually necessary here? We should always have `min(cdf, ccdf) <= 1//2` but maybe there were some numerical issues that were fixed by the additional truncation to 1?

Reply:
OK, promoting should be fine (it's just a bit longer to write). It seems safer to keep `min`, as we have seen several numerical issues with p-values. But `clamp` is probably more convenient as it converts the result.

Reply:
I've seen these issues but I still wonder if  here in this method values above 1 should just be considered an upstream bug. And if hence it is safe to assume that values are valid given that `cdf` and `ccdf` are implemented correctly.

Reply:
Basically similar to how we don't clamp values in the other branches because we assume upstream packages ensure that they are between 0 and 1.

Reply:
Makes sense.

Reply:
But, of course, another alternative would be to clamp them to the unit interval in all branches - I just think it shouldn't be necessary if there are no bugs upstream.

I guess I could be convinced otherwise if there's an example where we get invalid results even though there are no upstream bugs.

Reply:
Yeah I agree that when the upstream function is mathematically supposed to returned values in a given interval, it's more appropriate to avoid clamping so that we don't hide bugs. Cases where we need to clamp are when doing computations ourselves or when calling BLAS, as it allows for some roundoff error (and even in these cases, it would probably be safer to only allow for some tolerance when clamping, as we recently discussed elsewhere).

Reply:
It only applies to continuous distributions though, for discrete ones the values might not sum to 1 but something larger since it's P(X <= x) and P(X >= x). I guess generally the discrete case is somewhat broken since `ccdf(d, x - 1)` is just a heuristic. Ideally we want a function that returns P(X >= x) directly instead of working with P(X > x), possibly using an approach such as https://github.com/JuliaStats/Distributions.jl/issues/1483.

Reply:
So I'll revert the discrete case.